lib/ecto/adapters/mysql/connection.ex

```diff
@@ -0,0 +1,502 @@
+        Enum.map_join(fields, ", ", &expr(&1, sources))
+    end
+
+    defp from(sources) do
+      {table, name, _model} = elem(sources, 0)
+      "FROM #{quote_name(table)} AS #{name}"
+    end
+
+    defp join([], _sources), do: nil
+    defp join(joins, sources) do
+      Enum.map_join(joins, " ", fn
+        %JoinExpr{on: %QueryExpr{expr: expr}, qual: qual, ix: ix} ->
+          {table, name, _model} = elem(sources, ix)
+
+          on   = expr(expr, sources)
+          qual = join_qual(qual)
+
+          "#{qual} JOIN #{quote_name(table)} AS #{name} ON " <> on
+      end)
+    end
+
+    defp join_qual(:inner), do: "INNER"
+    defp join_qual(:left),  do: "LEFT OUTER"
+    defp join_qual(:right), do: "RIGHT OUTER"
+    defp join_qual(:full),  do: "FULL OUTER"
+
+    defp where(wheres, sources) do
+      boolean("WHERE", wheres, sources)
+    end
+
+    defp having(havings, sources) do
+      boolean("HAVING", havings, sources)
+    end
+
+    defp group_by(group_bys, sources) do
+      exprs =
+        Enum.map_join(group_bys, ", ", fn
+          %QueryExpr{expr: expr} ->
+            Enum.map_join(expr, ", ", &expr(&1, sources))
+        end)
+
+      case exprs do
+        "" -> nil
+        _  -> "GROUP BY " <> exprs
+      end
+    end
+
+    defp order_by(order_bys, sources) do
+      exprs =
+        Enum.map_join(order_bys, ", ", fn
+          %QueryExpr{expr: expr} ->
+            Enum.map_join(expr, ", ", &order_by_expr(&1, sources))
+        end)
+
+      case exprs do
+        "" -> nil
+        _  -> "ORDER BY " <> exprs
+      end
+    end
+
+    defp order_by_expr({dir, expr}, sources) do
+      str = expr(expr, sources)
+      case dir do
+        :asc  -> str
+        :desc -> str <> " DESC"
+      end
+    end
+
+    defp limit(nil, _sources), do: nil
+    defp limit(%Ecto.Query.QueryExpr{expr: expr}, sources) do
+      "LIMIT " <> expr(expr, sources)
+    end
+
+    defp offset(nil, _sources), do: nil
+    defp offset(%Ecto.Query.QueryExpr{expr: expr}, sources) do
+      "OFFSET " <> expr(expr, sources)
+    end
+
+    defp lock(nil), do: nil
+    defp lock(lock_clause), do: lock_clause
+
+    defp boolean(_name, [], _sources), do: nil
+    defp boolean(name, query_exprs, sources) do
+      name <> " " <>
+        Enum.map_join(query_exprs, " AND ", fn
+          %QueryExpr{expr: expr} ->
+            "(" <> expr(expr, sources) <> ")"
+        end)
+    end
+
+    defp expr({:^, [], [ix]}, _sources) do
+      "$#{ix+1}"
+    end
+
+    defp expr({{:., _, [{:&, _, [idx]}, field]}, _, []}, sources) when is_atom(field) do
+      {_, name, _} = elem(sources, idx)
+      "#{name}.#{quote_name(field)}"
+    end
+
+    defp expr({:&, _, [idx]}, sources) do
+      {_table, name, model} = elem(sources, idx)
+      fields = model.__schema__(:fields)
+      Enum.map_join(fields, ", ", &"#{name}.#{quote_name(&1)}")
+    end
+
+    defp expr({:in, _, [_left, []]}, _sources) do
+      "false"
```

Comment:
We need to check if this works for MySQL. If we do `SELECT ... FROM ... WHERE FALSE` does it return no results? For example, Rails was used to do `1=0` here for MySQL for some reason.